File: Cocinero.py
import threading
import time
import random

class Cocinero(threading.Thread):
    """Hilo que representa un cocinero que procesa pedidos concurrentemente.""" 

    def __init__(self, nombre, pedidos, log_file, lock):
         
        """Inicializa un cocinero.
        :param nombre: Nombre del cocinero.
        :param pedidos: Lista compartida de pedidos.
        :param log_file: Archivo donde se escribe el log.
        :param lock: Objeto para sincronizar acceso a la lista y log."""
        super().__init__()
        self.nombre = nombre
        self.pedidos = pedidos
        self.log_file = log_file
        self.lock = lock
    
    def run(self):
        """Ejecuta el hilo: toma pedidos, los prepara y los registra."""
        while True:
            pedido_actual = None
            # Acceso sincronizado a la lista de pedidos
            with self.lock:
                if not self.pedidos:
                    break
                pedido_actual = self.pedidos.pop(0)
            
            print(f"-> {self.nombre} está preparando el {pedido_actual.nombre_plato}")
            time.sleep(random.uniform(1, 3))# Simula tiempo de preparación
            
            mensaje = f"Completado: {self.nombre} terminó el {pedido_actual}"
            print(mensaje)
            # Acceso sincronizado al archivo de log
            with self.lock:
                self.log_file.write(mensaje + "\n")
                self.log_file.flush()

File: Cocinero.java
import java.io.PrintWriter;
import java.util.List;
import java.util.Random;
/**
 * Hilo que representa un cocinero que procesa pedidos de forma concurrente.
 */
public class Cocinero extends Thread {
    private final String nombre;// Nombre del cocinero
    private final List<Pedido> pedidos;// Lista compartida de pedidos
    private final PrintWriter log;// Registro de acciones completadas
    private final Object lock;// Objeto para sincronización
    private static final Random random = new Random();

    /**
     * Crea un cocinero.
     * @param nombre nombre del cocinero
     * @param pedidos lista compartida de pedidos
     * @param log salida donde se escriben logs
     * @param lock objeto para sincronizar acceso a la lista
     */
    public Cocinero(String nombre, List<Pedido> pedidos, PrintWriter log, Object lock) {
        this.nombre = nombre;
        this.pedidos = pedidos;
        this.log = log;
        this.lock = lock;
    }

    /** Ejecuta el hilo: toma pedidos, los prepara y los registra. */
    @Override
    public void run() {
        while (true) {
            Pedido pedidoActual;
            
            synchronized (lock) {
                if (pedidos.isEmpty()) break;
                pedidoActual = pedidos.remove(0);
            }

            System.out.println("-> " + nombre + " está preparando el " + pedidoActual.getNombrePlato());

            try {
                Thread.sleep(random.nextInt(2000) + 1000);//Dejamos solo el Thread ya que el catch se enfoca solo en ver la interrupcion de este.
              } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                System.err.println("El cocinero " + nombre + " fue interrumpido.");
                break;
            }   
                            
            String mensaje = "Completado: " + nombre + " terminó el " + pedidoActual;
                System.out.println(mensaje);

             synchronized (lock) {
                    log.println(mensaje);
                    log.flush();
            }
        }
    }
}

